[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71269

Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

dzdang committed:
[Quant] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case

Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1121,6 +1139,36 @@ static Tensor select_sparse(const Tensor& self, int64_t dim, int64_t index) {
+  auto zero_points = temp->zero_points();
+  if (dim == axis) {
+    // Compute scales&zps for sub-tensor
+    is_select ? scales = scales.select(0, start) : scales = scales.slice(0, start, end, step);
+    is_select ? zero_points = zero_points.select(0, start) : zero_points = zero_points.slice(0, start, end, step);
```
can select be expressed with slice?

jerryzh168 replied:
also, probably clearer if we do:
```
scales = is_select ? ... : ...;
```
if we need to do this

dzdang replied:
> can select be expressed with slice?

i think so, let me think. slice would maintain the dimension of the scales&zp to be 1 but select would just make it a scalar but that shouldn't affect the rest of the code

jerryzh168 commented on aten/src/ATen/native/quantized/Copy.cpp:
```diff
@@ -22,11 +27,19 @@ Tensor& quantized_copy_from_float_cpu_(Tensor& self, const Tensor& src) {
+                                         self.q_per_channel_zero_points(),
+                                         self.q_per_channel_axis());
+    } else {
+      float* src_data = src.data_ptr<float>();
+      scalar_t* self_data = self.data_ptr<scalar_t>();
+      for (const auto i : c10::irange(self.numel())) {
+        auto sc = self.q_scale();
+        auto test = self.q_zero_point();
+        self_data[i] = quantize_val<scalar_t>(
+            self.q_scale(), self.q_zero_point(), src_data[i]);
```
can we call the quantize_tensor_per_tensor_affine here as well? (can be a separate PR)

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -257,6 +257,45 @@ def _test_qtensor_sub_byte(self, rows, cols, dtype, elements_per_byte, expected_
+                qt1[:1] = t2[:1]
+                self.assertEqual(qt1[:1], qt2[:1])
+                qt1[:] = t2[:]
+                self.assertEqual(qt1[:], qt2[:])
```
should we add a test for non-contiguous tensor slicing as well? 
we can use assertRaises: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L274

jerryzh168 reviewed (approved):
Looks great, thanks for all the detailed comments and tests!

dzdang committed:
Update on "[Quant] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

dzdang committed:
Update on "[Quant] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][new_features] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

dzdang committed:
Update on "[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

dzdang committed:
Update on "[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

dzdang committed:
Update on "[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

dzdang committed:
Update on "[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).

dzdang committed:
Update on "[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][improvement] Enabled slicing on per-channel quantized tensors (support is limited to the a contiguous sliced tensor); added correpsonding test case"


Summary: Support for slicing of per-channel quantized tensors was added; however, this feature is only available to contiguous sliced tensors, e.g., qt[1, :], but not qt[:, 0] for a 2x2 qt. The latter case would trigger an exception.

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestQuantizedTensor.test_qtensor_channel_float_assignment
```
to execute the corresponding test case.

Differential Revision: [D33943512](https://our.internmc.facebook.com/intern/diff/D33943512)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33943512).
